cmd/gaia/init/testnet.go

```diff
@@ -182,67 +201,115 @@ func testnetWithConfig(config *cfg.Config, cdc *codec.Codec) error {
 			return err
 		}
 
-		// Gather gentxs folder
+		// gather gentxs folder
 		err = writeFile(fmt.Sprintf("%v.json", nodeDirName), gentxsDir, txBytes)
 		if err != nil {
 			_ = os.RemoveAll(outDir)
 			return err
 		}
 	}
 
-	// Generate empty genesis.json
+	if err := initGenFiles(cdc, chainID, accs, genFiles); err != nil {
+		return err
+	}
+
+	if err := collectGenFiles(cdc, config, chainID, monikers, nodeIDs, valPubKeys); err != nil {
+		return err
+	}
+
+	fmt.Printf("Successfully initialized %v node directories\n", viper.GetInt(nValidators))
+	return nil
+}
+
+func initGenFiles(
+	cdc *codec.Codec, chainID string, accs []app.GenesisAccount, genFiles []string,
+) error {
+
+	numValidators := viper.GetInt(nValidators)
 	appGenState := app.NewDefaultGenesisState()
 	appGenState.Accounts = accs
+
 	appGenStateJSON, err := codec.MarshalJSONIndent(cdc, appGenState)
 	if err != nil {
 		return err
 	}
+
 	genDoc := types.GenesisDoc{
 		ChainID:    chainID,
 		AppState:   appGenStateJSON,
 		Validators: nil,
 	}
 
-	// Save all genesis.json files
+	// generate empty genesis files for each validator and save
 	for i := 0; i < numValidators; i++ {
 		if err := genDoc.SaveAs(genFiles[i]); err != nil {
 			return err
 		}
 	}
 
+	return nil
+}
+
+func collectGenFiles(
+	cdc *codec.Codec, config *cfg.Config, chainID string,
+	monikers, nodeIDs []string, valPubKeys []crypto.PubKey,
+) error {
+
+	outDir := viper.GetString(outputDir)
```

Comment:
Please pass these two variables as arguments to the function.

Reply:
@alessio why? The function sig is already super long. Under non circumstance would they differ.

Reply:
I'd avoid spread `viper` pollution across the module. After, it's just a couple of extra arguments, aren't they?